### PR TITLE
Implement `const fn new` for `Float`

### PR DESCRIPTION
**Pull request by azriel91**
## Description

~~Change `Float(FloatBase)` to `Float(pub FloatBase)`, so that we can construct constants.~~

Implemented `const fn new(value: <f32/64>)` for `Float` based on feature flags. This allows us to declare `Float` constants.

```rust
const FIELD: Float = Float::new(123.);
```

## Modifications

* Make `Float` inner field public.

## PR Checklist

By placing an x in the boxes I certify that I have:

- [x] Ran `cargo test --all` locally if this modified any rs files.
- [x] Ran `cargo +stable fmt --all` locally if this modified any rs files.
- **n/a** Updated the content of the book if this PR would make the book outdated.
- [x] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- **n/a** Added unit tests for new APIs if any were added in this PR.
- [x] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.

### CODE REVIEW

**Comment by jojolepro** [comment]
This leaks the inner type and goes against the encapsulation provided by Float which has for goal to prevent the real type from ever being publicly visible. I would suggest to not do it this way.
Being able to do float.0 + (some_var as f32) would break depending on the feature flag being used or not.

**Comment by azriel91** [comment]
ah, wonder if it works if `Float::new` is made into a `const fn`

**Comment by azriel91** [comment]
How's it looking now?

File: amethyst_core/src/float.rs [hunk 1]
```diff
@@ -32,11 +32,22 @@ type FloatBase = f32;
 pub struct Float(FloatBase);
 
 impl Float {
+    /// Returns a new `Float` from a `f64`.
+    pub const fn from_f64(value: f64) -> Self {
+        Float(value)
+    }
+
+    /// Returns a new `Float` from a `f64`.
```
Comment (jojolepro) [inline amethyst_core/src/float.rs]:
```suggestion
    /// Returns a new `Float` from a `f32` using a constant function.
```

File: amethyst_core/src/float.rs [hunk 2]
```diff
@@ -32,11 +32,22 @@ type FloatBase = f32;
 pub struct Float(FloatBase);
 
 impl Float {
+    /// Returns a new `Float` from a `f64`.
```
Comment (jojolepro) [inline amethyst_core/src/float.rs]:
```suggestion
    /// Returns a new `Float` from a `f64` using a constant function.
```

File: amethyst_core/src/float.rs (hunk 1 above)
Comment (Frizi) [reply amethyst_core/src/float.rs]: Same as above. Not necessarily const always.

Comment (Frizi) [reply amethyst_core/src/float.rs]: Although in this case it should be `f32` indeed :D

**Comment by Frizi** [comment]
We should also make reverse operation `const`, like `into_f32` and `into_f64`. That way you can do whole const expressions in chosen precision without ever leaking the inner type.

File: amethyst_core/src/float.rs (hunk 1 above)
Comment (azriel91) [reply amethyst_core/src/float.rs]: oops, fixed that typo!

**Comment by azriel91** [comment]
cool, that last one -- changing `as_f64` to `const` changed the impl to use `as f64` instead of `f64::from` -- as `from` is not `const`

**Pull request merged**